src/main/java/dev/blaauwendraad/masker/json/KeyMatcher.java

```diff
@@ -520,4 +578,73 @@ void addUpper(Byte b, PreInitTrieNode child) {
             childrenUpper.put(b, child);
         }
     }
+
+    /**
+     * A helper class for managing (radix) trie node and the sequence. On top of regular methods
+     * used for matching, also contains {@link #checkpoint()} and {@link #restore()} to allow
+     * repeated matching that would otherwise pollute the sequence.
+     */
+    static class StatefulRadixTrieNode {
+        private TrieNode node;
+        private int sequence;
+
+        private TrieNode checkPointNode;
+        private int checkPointSequence;
+
+        StatefulRadixTrieNode(TrieNode node) {
+            this.node = this.checkPointNode = node;
+            this.sequence = this.checkPointSequence = 0;
+        }
+
+        StatefulRadixTrieNode(StatefulRadixTrieNode node) {
+            this.node = this.checkPointNode = node.node;
+            this.sequence = this.checkPointSequence = node.sequence;
+        }
+
+        @Nullable
+        StatefulRadixTrieNode child(byte b) {
```

Comment:
`getChild` maybe? I think this one can also 1 sentence JavaDoc what `child()` means (it returns the trie node with the child currently set as "active node"? )